pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java

```diff
@@ -342,56 +344,101 @@ public CompletableFuture<Optional<BrokerLookupData>> assign(Optional<ServiceUnit
+            }
+            return getBrokerLookupData(owner, bundle);
+        });
+    }
+
+    private CompletableFuture<Optional<String>> getOwnerAsync(
+            ServiceUnitId serviceUnit, String bundle, boolean ownByLocalBrokerIfAbsent) {
+        return serviceUnitStateChannel.getOwnerAsync(bundle).thenCompose(broker -> {
+            // If the bundle not assign yet, select and publish assign event to channel.
+            if (broker.isEmpty()) {
+                CompletableFuture<Optional<String>> selectedBroker;
+                if (ownByLocalBrokerIfAbsent){
+                    String brokerId = this.brokerRegistry.getBrokerId();
+                    selectedBroker = CompletableFuture.completedFuture(Optional.of(brokerId));
+                } else {
+                    selectedBroker = this.selectAsync(serviceUnit);
+                }
+                return selectedBroker.thenCompose(brokerOpt -> {
+                    if (brokerOpt.isPresent()) {
+                        assignCounter.incrementSuccess();
+                        log.info("Selected new owner broker: {} for bundle: {}.", brokerOpt.get(), bundle);
+                        return serviceUnitStateChannel.publishAssignEventAsync(bundle, brokerOpt.get())
+                                .thenApply(Optional::of);
+                    } else {
+                        throw new IllegalStateException(
+                                "Failed to select the new owner broker for bundle: " + bundle);
                     }
-                    assignCounter.incrementSkip();
-                    // Already assigned, return it.
-                    return CompletableFuture.completedFuture(broker);
                 });
             }
+            assignCounter.incrementSkip();
+            // Already assigned, return it.
+            return CompletableFuture.completedFuture(broker);
+        });
+    }
 
-            return owner.thenCompose(broker -> {
-                if (broker.isEmpty()) {
-                    String errorMsg = String.format(
-                            "Failed to get or assign the owner for bundle:%s", bundle);
-                    log.error(errorMsg);
-                    throw new IllegalStateException(errorMsg);
-                }
-                return CompletableFuture.completedFuture(broker.get());
-            }).thenCompose(broker -> this.getBrokerRegistry().lookupAsync(broker).thenCompose(brokerLookupData -> {
-                if (brokerLookupData.isEmpty()) {
-                    String errorMsg = String.format(
-                            "Failed to look up a broker registry:%s for bundle:%s", broker, bundle);
-                    log.error(errorMsg);
-                    throw new IllegalStateException(errorMsg);
-                }
-                return CompletableFuture.completedFuture(brokerLookupData);
-            }));
+    private CompletableFuture<Optional<BrokerLookupData>> getBrokerLookupData(
```

Comment:
yes, the same idea with @mattisonchao

Reply:
The `serviceUnitStateChannel.getChannelOwnerAsync()` will return `Optional`, so we can change this method.